<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [2]:
# load in the appropriate path to the file you want to load in
rockfile = "/data/rock.csv"

In [32]:
# Load the data.
df = pd.read_csv('../../data/rock.csv', names=col_names)

In [31]:
# Look at the information regarding its columns.
df.columns = ['This', 'That', 'Something Else', 'Word', 'Word2', 'Year', 'Year3', 'Year4']

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [5]:
# Change the column names when loading the '.csv':

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [6]:
# Change the column names using the `.rename()` function.

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [7]:
# Replace the column names by reassigning the `.columns` attribute.

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [33]:
# Show records where df['release'] is null
df['Release'].isnull().sum()

577

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [41]:
# Replace release nulls with 0
df['Release'].fillna(value=0, inplace=True)

In [46]:
cat_columns = df.select_dtypes(include=np.object).columns.tolist()

In [49]:
df[cat_columns] = df[cat_columns].fillna(df[cat_columns].max())

In [36]:
mask

0       False
1       False
2        True
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11       True
12      False
13      False
14       True
15      False
16      False
17       True
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25       True
26       True
27       True
28      False
29      False
        ...  
2201    False
2202    False
2203    False
2204     True
2205    False
2206     True
2207     True
2208    False
2209     True
2210    False
2211    False
2212    False
2213    False
2214    False
2215    False
2216     True
2217     True
2218    False
2219     True
2220    False
2221    False
2222     True
2223     True
2224    False
2225    False
2226     True
2227    False
2228    False
2229    False
2230    False
Name: Release, Length: 2231, dtype: bool

#### 4.B Verify that `release` contains no null values.

In [50]:
# A:
df['Release'].isnull().sum()

0

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [51]:
# A:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2231 entries, 0 to 2230
Data columns (total 8 columns):
Song         2231 non-null object
Artist       2231 non-null object
Release      2231 non-null object
Combined     2231 non-null object
First        2231 non-null object
Year         2231 non-null object
PlayCount    2231 non-null object
FG           2231 non-null object
dtypes: object(8)
memory usage: 139.5+ KB


In [52]:
df['Release'].unique()

array(['Release Year', '1982', 0, '1981', '1980', '1975', '2000', '2002',
       '1992', '1985', '1993', '1976', '1995', '1979', '1984', '1977',
       '1990', '1986', '1974', '2014', '1987', '1973', '2001', '1989',
       '1997', '1971', '1972', '1994', '1970', '1966', '1965', '1983',
       '1955', '1978', '1969', '1999', '1968', '1988', '1962', '2007',
       '1967', '1958', '1071', '1996', '1991', '2005', '2011', '2004',
       '2012', '2003', '1998', '2008', '1964', '2013', '2006',
       'SONGFACTS.COM', '1963', '1961'], dtype=object)

In [54]:
mask = (df.Release == 'Release Year') | (df.Release == 'SONGFACTS.COM')

In [58]:
df.loc[mask, 'Release'] = 0

In [60]:
df.isnull().sum() > 0

Song         False
Artist       False
Release      False
Combined     False
First        False
Year         False
PlayCount    False
FG           False
dtype: bool

In [ ]:
df.loc[:, df.isnull().sum() > 0]

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [12]:
# A:
df['Release Year'].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM', '1963',
       '1961'], dtype=object)

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [16]:
# A:
df.loc[df['Release Year'] == 'SONGFACTS.COM', 'Release Year'] = 0

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [14]:
# A:

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [61]:
# A:
df['Release'].describe()

count     2231
unique      56
top          0
freq       579
Name: Release, dtype: int64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [19]:
# A:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    2230 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


In [62]:
df['Release'].unique()

array([0, '1982', '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', '1963', '1961'],
      dtype=object)

In [65]:
df.dtypes

Song         object
Artist       object
Release      object
Combined     object
First        object
Year         object
PlayCount    object
FG           object
dtype: object

In [68]:
2**16

65536

In [69]:
df['Release'] = df['Release'].astype(np.int16)

In [70]:
df['Release'].dtype

dtype('int16')

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [20]:
# A:
df['Release Year'] = df['Release Year'].astype(np.int16)

In [21]:
df.columns

Index(['Song Clean', 'ARTIST CLEAN', 'Release Year', 'COMBINED', 'First?',
       'Year?', 'PlayCount', 'F*G'],
      dtype='object')

In [24]:
def prints_info(row):
    return row['Song Clean'], row['ARTIST CLEAN'], row['Release Year'] < 1970

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [25]:
# A:
df.head().apply(prints_info, axis=1)

0          (Caught Up in You, .38 Special, False)
1               (Fantasy Girl, .38 Special, True)
2           (Hold On Loosely, .38 Special, False)
3    (Rockin' Into the Night, .38 Special, False)
4                (Art For Arts Sake, 10cc, False)
dtype: object

In [73]:
def some_function(val):
    return val**2

In [74]:
df['Release'].head().apply(some_function)

0          0
1    3928324
2          0
3    3924361
4    3920400
Name: Release, dtype: int64